Indicators/CorrelationIndicator.cs

```diff
@@ -0,0 +1,211 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using QuantConnect.Data.Market;
+using MathNet.Numerics.Statistics;
+using System.Linq;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The Correlation Indicator is a valuable tool in technical analysis, designed to quantify the degree of 
+    /// relationship between the price movements of a target security (e.g., a stock or ETF) and a reference 
+    /// market index. It measures how closely the target’s price changes are aligned with the fluctuations of 
+    /// the index over a specific period of time, providing insights into the target’s susceptibility to market 
+    /// movements.
+    /// A positive correlation indicates that the target tends to move in the same direction as the market index, 
+    /// while a negative correlation suggests an inverse relationship. A correlation close to 0 implies a weak or 
+    /// no linear relationship.
+    /// Commonly, the SPX index is employed as the benchmark for the overall market when calculating correlation, 
+    /// ensuring a consistent and reliable reference point. This helps traders and investors make informed decisions 
+    /// regarding the risk and behavior of the target security in relation to market trends.
+    /// </summary>
+
+    public class CorrelationIndicator : TradeBarIndicator, IIndicatorWarmUpPeriodProvider
```

Comment:
- Think the indicator name should just be `Correlation`, no need for the Indicator part 👌. 
- Missing to add the helper method in the `QCAlgorith.Indicators.cs`
- I actually think the indicator base type doesn't really need to be `TradeBar`, the input doesn't even need to be a Bar, just needs a `Value` which is part of the basic BaseData? Suggest looking at other generic indicators & how they handle this, like SMA

Reply:
Yes, agree that the class should be called `Correlation`, but we are using method: `Correlation` from `MathNet.Numerics.Statistics`.
For avoid writing a full namespace path and future mistakes decided to change the name of Indicator to `CorrelationIndicator`.
I can use aliases:
```
using Correlation = MathNet.Numerics.Statistics.Correlation;
using CorrelationIndicator = QuantConnect.Indicators.Correlation;
```
But I'm not sure if it will be a good practice.
I'm open to any suggestion.

Reply:
I tried to change `TradeBar` to `IndicatorDataPoint`, but I noticed that there is a risk that the indicator will not get a Symbol Name.
In the correlation indicator important is to separate 2 data sources using Symbol Name.

Reply:
I think the full namespace is ok in this case, it is used just twice in the same location.
Also, could handle indicator data points without symbol by throwing an exception. It would be just like passing a bar with a symbol that is neither the target not the reference.

Reply:
We can avoid always referencing the Numerics version as its not likely many people using it.